File: ObserverPattern.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.random.RandomGenerator;

interface Observer {
    void update(Observable observable);
}

interface Observable {
    void addObserver(Observer observer);
    void removeObserver(Observer observer);
    void notifyObservers();
    void printItems();
}
class Item{
    int id;
    Item(){Random rand = new Random();this.id = rand.nextInt((int)1e5);}
    int getPrice(){return 0;}
    @Override
    public String toString(){
        return ""+this.id;
    }
}

class ShoppingCart implements Observable {
    
    private List<Item> items;
    private List<Observer> observers;

    public ShoppingCart() {
        this.items = new ArrayList<>();
        this.observers = new ArrayList<>();
    }

    public void addItem(Item item) {
        items.add(item);
        notifyObservers();
    }

    public void removeItem(Item item) {
        items.remove(item);
        notifyObservers();
    }

    public double getTotalPrice() {
        double totalPrice = 0.0;
        for (Item item : items) {
            totalPrice += item.getPrice();
        }
        return totalPrice;
    }

    @Override
    public void addObserver(Observer observer) {
        observers.add(observer);
    }

    @Override
    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers() {
        for (Observer observer : observers) {
            observer.update(this);
        }
        printItems();
        System.out.println("--------------");
    }
    @Override
    public void printItems(){
        for(Item it:items){
            System.out.println(it);
        }
    }
    
}
public class ObserverPattern{
    public static void main(String[] args) {
        ShoppingCart sCart = new ShoppingCart();
        for(int i = 0;i<5;i++) sCart.addItem(new Item());
    }
}

File: ObserverPattern.py
class Observer:
    def update(self, value):
        pass

class Observable:
    def __init__(self):
        self.observers = []

    def add_observer(self, observer):
        self.observers.append(observer)

    def remove_observer(self, observer):
        self.observers.remove(observer)

    def notify_observers(self, value):
        for observer in self.observers:
            observer.update(value)

class Data(Observable):
    def __init__(self):
        super().__init__()
        self._value = 0

    @property
    def value(self):
        return self._value

    @value.setter
    def value(self, value):
        self._value = value
        self.notify_observers(value)

class DataObserver(Observer):
    def update(self, value):
        print(f"Value updated: {value}")

data = Data()
observer = DataObserver()
data.add_observer(observer)
data.value = 10
data.value = 20
